File: src/themes/registry.js
// Theme registry — lazy-loaded catalog of manifests.
//
// Every entry is a function that returns a dynamic import, so vite code-splits
// each theme (and its background component + assets) into its own chunk. The
// classic themes are bundled together because they're tiny (tokens only, no
// background) and the user almost certainly wants one of them as fallback.
//
// Legacy-id mapping lives here so the provider can silently upgrade old
// localStorage values without the user noticing. See LEGACY_THEME_MAP.

/** @type {Record<string, () => Promise<{default: import('./types.js').ThemeManifest}>>} */
const CATALOG = {
  // Four refreshed themes based on the 2026-04 design handoff. Graphite is
  // the "standard" default (from STANDART.HTML mockup); the others are each
  // paired with a dedicated Background component.
  'classic-graphite': () => import('./catalog/classic/graphiteManifest.js'),
  'classic-light':    () => import('./catalog/classic/lightManifest.js'),
  'classic-matrix':   () => import('./catalog/classic/matrixManifest.js'),
  'sakura-zen':       () => import('./catalog/atmospheric/sakuraZenManifest.js'),
};

/**
 * Legacy → current id map. Applied once when ThemeProvider reads localStorage
 * so a user who installed the old build keeps their preference.
 *
 * Historical ids come from two sources:
 *  - App.jsx: `obsidian`, `sakura`, `matrix` (via <html data-theme>)
 *  - themeManager.js: `sakura_zen`, `aurora_flow`, `retro_synth`, `matrix`,
 *    `obsidian`, `none`, plus its own legacy list (`aurora`, `stellar`, …)
 */
export const LEGACY_THEME_MAP = Object.freeze({
  // Retired Obsidian (classic-dark) and every pre-refresh dark theme fall back
  // to Graphite — the 2026-04 default.
  'classic-dark': 'classic-graphite',
  obsidian:       'classic-graphite',
  sakura:         'classic-graphite',
  matrix:         'classic-matrix',

  // From the (dead) themeManager catalog
  sakura_zen:  'sakura-zen',
  aurora_flow: 'classic-graphite',
  retro_synth: 'classic-graphite',
  none:        'classic-graphite',

  // From themeManager's own legacy map
  aurora:  'classic-graphite',
  stellar: 'classic-graphite',
  retro:   'classic-graphite',
  japan:   'classic-graphite',
  abyss:   'classic-graphite',
  draft:   'classic-graphite'
});

export const DEFAULT_THEME_ID = 'classic-graphite';

export function listThemeIds() {
  return Object.keys(CATALOG);
}

export function hasTheme(id) {
  return Object.prototype.hasOwnProperty.call(CATALOG, id);
}

/**
 * Resolve any historical id to a currently-valid manifest id. Idempotent:
 * passing an already-valid id returns it unchanged.
 */
export function canonicalizeThemeId(rawId) {
  if (!rawId || typeof rawId !== 'string') return DEFAULT_THEME_ID;
  if (hasTheme(rawId)) return rawId;
  const mapped = LEGACY_THEME_MAP[rawId];
  if (mapped && hasTheme(mapped)) return mapped;
  return DEFAULT_THEME_ID;
}

/**
 * Load a manifest by id. Falls back to DEFAULT_THEME_ID on any failure
 * so the UI is never stuck on a broken theme.
 *
 * @returns {Promise<import('./types.js').ThemeManifest>}
 */
export async function loadThemeManifest(id) {
  const canonical = canonicalizeThemeId(id);
  const loader = CATALOG[canonical] || CATALOG[DEFAULT_THEME_ID];
  try {
    const mod = await loader();
    return mod.default;
  } catch (err) {
    try { console.warn(`[themes] failed to load "${canonical}"`, err); } catch (_) {}
    if (canonical !== DEFAULT_THEME_ID) {
      // Try the fallback once, and give up if that fails too.
      const fb = await CATALOG[DEFAULT_THEME_ID]();
      return fb.default;
    }
    throw err;
  }
}

File: lib/themes/registry.dart
// Theme registry — Flutter port of `git_push/src/themes/registry.js`.
//
// JS code-splits each manifest behind a dynamic `import()`. On Flutter every
// manifest is part of the same bundle (no chunk-splitting on the web build
// today; all themes are tiny anyway), so the catalog is a simple `const`
// map. That also means `loadThemeManifest` is synchronous — no `Future`,
// no error-recovery dance.
//
// `canonicalizeThemeId` keeps doing the same job as JS: any historical id
// from older builds (`obsidian`, `sakura`, `aurora`, …) maps to the closest
// 2026-04 equivalent so users who upgrade don't land on a broken theme.

import 'manifest.dart';
import 'catalog/atmospheric/sakura_zen_manifest.dart';
import 'catalog/classic/graphite_manifest.dart';
import 'catalog/classic/light_manifest.dart';
import 'catalog/classic/matrix_manifest.dart';

/// Stable id of the theme used when the user has never picked one (or when
/// the persisted id can't be resolved). Matches JS `DEFAULT_THEME_ID`.
const String defaultThemeId = 'classic-graphite';

/// SharedPreferences key holding the user's persisted choice. Same value
/// as the JS localStorage key so a cross-platform user shares preferences.
const String themePrefsKey = 'orbits_theme';

/// id → manifest. Ordering here drives ordering in the picker. Mirrors the
/// JS `CATALOG` object (with an unhelpful map type in JS, here typed for us).
///
/// Const because every manifest is itself const — each one's `background:`
/// is a top-level function tear-off (`_buildGraphite`, …) which Dart treats
/// as a compile-time constant. The function bodies still close over the
/// matching `graphiteManifest` (etc.) without breaking const evaluation.
const Map<String, ThemeManifest> themeCatalog = <String, ThemeManifest>{
  'classic-graphite': graphiteManifest,
  'classic-light': paperManifest,
  'classic-matrix': matrixManifest,
  'sakura-zen': sakuraZenManifest,
};

/// Legacy → current id mapping. Applied once on read so a user with an old
/// `localStorage.orbits_theme = 'obsidian'` keeps a working theme.
///
/// Sources of legacy ids (per the JS `registry.js` audit):
///  - `App.jsx`         — `obsidian`, `sakura`, `matrix`
///  - `themeManager.js` — `sakura_zen`, `aurora_flow`, `retro_synth`,
///                        `matrix`, `obsidian`, `none`
///  - themeManager's own legacy list — `aurora`, `stellar`, `retro`, `japan`,
///    `abyss`, `draft`
const Map<String, String> legacyThemeMap = <String, String>{
  // Retired pre-refresh dark themes → Graphite (the 2026-04 default).
  'classic-dark': 'classic-graphite',
  'obsidian': 'classic-graphite',
  'sakura': 'classic-graphite',
  'matrix': 'classic-matrix',

  // From the (now-dead) themeManager catalog.
  'sakura_zen': 'sakura-zen',
  'aurora_flow': 'classic-graphite',
  'retro_synth': 'classic-graphite',
  'none': 'classic-graphite',

  // From themeManager's own legacy fallbacks.
  'aurora': 'classic-graphite',
  'stellar': 'classic-graphite',
  'retro': 'classic-graphite',
  'japan': 'classic-graphite',
  'abyss': 'classic-graphite',
  'draft': 'classic-graphite',
};

/// All currently-valid theme ids in catalog order.
List<String> listThemeIds() => themeCatalog.keys.toList(growable: false);

/// `true` if [id] is a *current* (not legacy) theme.
bool hasTheme(String id) => themeCatalog.containsKey(id);

/// Resolve any historical id to a current manifest id. Idempotent: passing
/// an already-valid id returns it unchanged. Empty/unknown → [defaultThemeId].
String canonicalizeThemeId(String? rawId) {
  if (rawId == null || rawId.isEmpty) return defaultThemeId;
  if (hasTheme(rawId)) return rawId;
  final mapped = legacyThemeMap[rawId];
  if (mapped != null && hasTheme(mapped)) return mapped;
  return defaultThemeId;
}

/// Load a manifest by id. Falls back to [defaultThemeId] if the id resolves
/// to nothing — there's no scenario where we hand back null and let the
/// caller crash.
ThemeManifest loadThemeManifest(String? id) {
  final canonical = canonicalizeThemeId(id);
  return themeCatalog[canonical] ?? themeCatalog[defaultThemeId]!;
}
